doc/manpages/srptool.1
.TH srptool 1 "21 Jan 2012" "3.0.12" "User Commands"
.\"
.\"  DO NOT EDIT THIS FILE   (srptool-args.man)
.\"  
.\"  It has been AutoGen-ed  January 21, 2012 at 11:32:42 AM by AutoGen 5.12
.\"  From the definitions    ../../src/srptool-args.def
.\"  and the template file   agman-cmd.tpl
.\"
.SH NAME
srptool \- GnuTLS SRP tool
.SH SYNOPSIS
.B srptool
.\" Mixture of short (flag) options and long options
.RB [ \-\fIflag\fP " [\fIvalue\fP]]... [" \-\-\fIopt\-name\fP " [[=| ]\fIvalue\fP]]..." 
.PP
Operands and options may be intermixed.  They will be reordered.
.SH "DESCRIPTION"
Simple program that emulates the programs in the Stanford SRP (Secure
Remote Password) libraries using GnuTLS.  It is intended for use in  places
where you don't expect SRP authentication to be the used for system users.
In  brief,  to use SRP you need to create two files. These are the password
file that holds the users and the verifiers associated with  them  and  the
configuration file to hold the group parameters (called tpasswd.conf).
.SH "OPTIONS"
.TP
.BR \-d " \fInumber\fP, " \-\-debug "=" \fInumber\fP
Enable debugging..
This option takes an integer number as its argument.
The value of \fInumber\fP is constrained to being:
.in +4
.nf
.na
in the range  0 through 9999
.fi
.in -4
.sp
Specifies the debug level.
.TP
.BR \-i ", " -\-index
specify the index of the group parameters in tpasswd.conf to use..
.sp
.TP
.BR \-u " \fIstring\fP, " \-\-username "=" \fIstring\fP
specify a username.
.sp
.TP
.BR \-p " \fIstring\fP, " \-\-passwd "=" \fIstring\fP
specify a password file..
.sp
.TP
.BR \-s " \fInumber\fP, " \-\-salt "=" \fInumber\fP
specify salt size for crypt algorithm..
This option takes an integer number as its argument.
.sp
.TP
.BR \-\-verify
just verify the password..
.sp
Verifies the password provided against the password file.
.TP
.BR \-v " \fIstring\fP, " \-\-passwd\-conf "=" \fIstring\fP
specify a password conf file..
.sp
Specify a filename or a PKCS #11 URL to read the CAs from.
.TP
.BR \-\-create\-conf "=\fIstring\fP"
Generate a password configuration file..
.sp
This generates a password configuration file (tpasswd.conf)
containing the required for TLS parameters.
.TP
.BR \-? , " \-\-help"
Display usage information and exit.
.TP
.BR \-! , " \-\-more-help"
Pass the extended usage information through a pager.
.TP
.BR \-v " [{\fIv|c|n\fP}]," " \-\-version" "[=\fI{v|c|n}\fP]"
Output version of program and exit.  The default mode is `v', a simple
version.  The `c' mode will print copyright information and `n' will
print the full copyright notice.
.SH EXAMPLES
To create \fItpasswd.conf\fP which holds the g and n values for SRP protocol
(generator and a large prime), run:
.br
.in +4
.nf
$ srptool \-\-create\-conf /etc/tpasswd.conf
.in -4
.fi
.sp
This command will create \fI/etc/tpasswd\fP and will add user 'test' (you
will also be prompted for a password). Verifiers are stored by default
in the way libsrp expects.
.br
.in +4
.nf
srptool \-\-passwd /etc/tpasswd \-\-passwd\-conf /etc/tpasswd.conf \-u test
.in -4
.fi
.sp
.sp
This command will check against a password. If the password matches
the one in \fI/etc/tpasswd\fP you will get an ok.
.br
.in +4
.nf
$ srptool \-\-passwd /etc/tpasswd \-\-passwd\-conf /etc/tpasswd.conf \-\-verify \-u test
.in -4
.fi
.SH "EXIT STATUS"
One of the following exit values will be returned:
.TP
.BR 0
Successful program execution.
.TP
.BR 1
The operation failed or the command syntax was not valid.
.SH "SEE ALSO"
    gnutls-cli-debug (1), gnutls-serv (1), srptool (1), psktool (1), certtool (1)
.SH "AUTHORS"
Nikos Mavrogiannopoulos and Simon Josefsson
.SH "COPYRIGHT"
Copyright (C) 2000-2012 Free Software Foundation all rights reserved.
This program is released under the terms of the GNU General Public License, version 3 or later.
.SH "BUGS"
Please send bug reports to: bug-gnutls@gnu.org
.SH "NOTES"
This manual page was \fIAutoGen\fP-erated from the \fBsrptool\fP
option definitions.


NEWS
GnuTLS NEWS -- History of user-visible changes.                -*- outline -*-
Copyright (C) 2000-2012 Free Software Foundation, Inc.
See the end for copying conditions.

* Version 3.0.13 (unreleased)

** ocsptool: Added --ask parameter, to verify a certificate's
status from an ocsp server.

** command line apps: Use gnu autogen (libopts) to parse command
line arguments and template files.

** libgnutls: SUITEB128 and SUITEB192 priority strings account
for the RFC6460 requirements.

** libgnutls: Added new security parameter GNUTLS_SEC_PARAM_LEGACY
to account for security level of 96-bits.

** libgnutls: In client side if server does not advertise any
known CAs and only a single certificate is set in the credentials,
sent that one.

** libgnutls: Added functions to parse authority key identifiers
when stored as a 'general name' and serial combo.

** libgnutls: Added function to force explicit reinitialization
of PKCS #11 modules. This is required on the child process after
a fork.

** API and ABI modifications:
gnutls_x509_crt_get_authority_key_gn_serial: Added
gnutls_x509_crl_get_authority_key_gn_serial: Added
gnutls_pkcs11_reinit: Added
gnutls_ecc_list: Added
gnutls_priority_certificate_type_list: Added
gnutls_priority_sign_list: Added
gnutls_priority_protocol_list: Added
gnutls_priority_compression_list: Added
gnutls_priority_ecc_curve_list: Added


* Version 3.0.12 (released 2012-01-20)

** libgnutls: Added OCSP support.
There is a new header file gnutls/ocsp.h and a set of new functions
under the gnutls_ocsp namespace.  Currently the functionality provided
is to parse and extract information from OCSP requests/responses, to
generate OCSP requests and to verify OCSP responses.  See the manual
for more information.  Run ./configure with --disable-ocsp to build
GnuTLS without OCSP support.

This work was sponsored by Smoothwall <http://smoothwall.net/>.

** ocsptool: Added new command line tool.
The tool can parse OCSP request/responses, generate OCSP requests and
verify OCSP responses.  See the manual for more information.

** certtool: --outder option now works for private
and public keys as well.

** libgnutls: Added error code GNUTLS_E_NO_PRIORITIES_WERE_SET
to warn when no or insufficient priorities were set.

** libgnutls: Corrected an alignment issue in ECDH
key generation which prevented some keys from being
correctly aligned in rare circumstances.

** libgnutls: Corrected memory leaks in DH parameter
generation and ecc_projective_check_point().

** libgnutls: Added gnutls_x509_dn_oid_name() to 
return a descriptive name of a DN OID.

** API and ABI modifications:
gnutls_pubkey_encrypt_data: Added
gnutls_x509_dn_oid_name: Added
gnutls_session_resumption_requested: Added
gnutls/ocsp.h: Added new header file.
gnutls_ocsp_print_formats_t: Added new type.
gnutls_ocsp_resp_status_t: Added new type.
gnutls_ocsp_cert_status_t: Added new type.
gnutls_x509_crl_reason_t: Added new type.
gnutls_ocsp_req_add_cert: Added.
gnutls_ocsp_req_add_cert_id: Added.
gnutls_ocsp_req_deinit: Added.
gnutls_ocsp_req_export: Added.
gnutls_ocsp_req_get_cert_id: Added.
gnutls_ocsp_req_get_extension: Added.
gnutls_ocsp_req_get_nonce: Added.
gnutls_ocsp_req_get_version: Added.
gnutls_ocsp_req_import: Added.
gnutls_ocsp_req_init: Added.
gnutls_ocsp_req_print: Added.
gnutls_ocsp_req_randomize_nonce: Added.
gnutls_ocsp_req_set_extension: Added.
gnutls_ocsp_req_set_nonce: Added.
gnutls_ocsp_resp_deinit: Added.
gnutls_ocsp_resp_export: Added.
gnutls_ocsp_resp_get_certs: Added.
gnutls_ocsp_resp_get_extension: Added.
gnutls_ocsp_resp_get_nonce: Added.
gnutls_ocsp_resp_get_produced: Added.
gnutls_ocsp_resp_get_responder: Added.
gnutls_ocsp_resp_get_response: Added.
gnutls_ocsp_resp_get_signature: Added.
gnutls_ocsp_resp_get_signature_algorithm: Added.
gnutls_ocsp_resp_get_single: Added.
gnutls_ocsp_resp_get_status: Added.
gnutls_ocsp_resp_get_version: Added.
gnutls_ocsp_resp_import: Added.
gnutls_ocsp_resp_init: Added.
gnutls_ocsp_resp_print: Added.
gnutls_ocsp_resp_verify: Added.

* Version 3.0.11 (released 2012-01-06)

** libgnutls: Corrected functionality of 
gnutls_record_get_direction(). Reported by Philip Allison.

** libgnutls: Provide less timing information when decoding
TLS/DTLS record packets. Patch by Nadhem Alfardan.

** API and ABI modifications:
No changes since last version.


* Version 3.0.10 (released 2012-01-04)

** gnutls-cli/serv: Set don't fragment bit in DTLS sessions
in Linux as well as in BSD.

** gnutls-cli: Fixed reading from windows terminals.

** libgnutls: When GNUTLS_OPENPGP_FMT_BASE64 is specified
the stream is assumed to be base64 encoded (previously
the encoding was auto-detected). This avoids a decoding
issue in windows systems.

** libgnutls: Corrected ciphersuite GNUTLS_ECDHE_PSK_AES_256_CBC_SHA384

** libgnutls: Added ciphersuites: GNUTLS_PSK_WITH_AES_256_GCM_SHA384
and GNUTLS_DHE_PSK_WITH_AES_256_GCM_SHA384.

** libgnutls: Added function gnutls_random_art() to convert 
fingerprints to images (currently ascii-art).

** libgnutls: Corrected bug in DSA private key parsing, which
prevented the verification of the key.

** API and ABI modifications:
gnutls_random_art: Added


* Version 3.0.9 (released 2011-12-13)

** certtool: Added new parameter --dh-info.

** certtool: -l option was overloaded so if combined with --priority
it will only list the ciphersuites that are enabled by the given
priority string.

** libgnutls: Added new priority string %SERVER_PRECEDENCE, which
changes the ciphersuite selection procedure. If specified the server
priorities will be used for selection instead of the client's.

** libgnutls: Optimizations in Diffie-Hellman parameters generation
and key exchange.

** libgnutls: When session tickets are negotiated and used in a
session, a server will not store that session data into its cache.

** libgnutls: Added the SECP192R1 curve.

** libgnutls: Added gnutls_priority_get_cipher_suite_index() to

src/srptool.c
/*
 * Copyright (C) 2001-2012 Free Software Foundation, Inc.
 *
 * This file is part of GnuTLS.
 *
 * GnuTLS is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * GnuTLS is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <gnutls/gnutls.h>
#include <gnutls/crypto.h>      /* for random */

#include <sys/types.h>
#include <sys/stat.h>

#ifndef _WIN32
#include <pwd.h>
#include <unistd.h>
#else
#include <windows.h>
#endif

/* Gnulib portability files. */
#include <getpass.h>
#include <minmax.h>
#include <progname.h>
#include <version-etc.h>

#include <gettext.h>
#include <srptool-args.h>

/* This may need some rewrite. A lot of stuff which should be here
 * are in the library, which is not good.
 */

int crypt_int (const char *username, const char *passwd, int salt,
               const char *tpasswd_conf, const char *tpasswd, int uindex);
static int read_conf_values (gnutls_datum_t * g, gnutls_datum_t * n,
                             char *str);
static int _verify_passwd_int (const char *username, const char *passwd,
                               char *verifier, const char *salt,
                               const gnutls_datum_t * g,
                               const gnutls_datum_t * n);

static void
print_num (const char *msg, const gnutls_datum_t * num)
{
  unsigned int i;

  printf ("%s:\t", msg);

  for (i = 0; i < num->size; i++)
    {
      if (i != 0 && i % 12 == 0)
        printf ("\n\t");
      else if (i != 0 && i != num->size)
        printf (":");
      printf ("%.2x", num->data[i]);
    }
  printf ("\n\n");

}

static int
generate_create_conf (const char *tpasswd_conf)
{
  FILE *fd;
  char line[5 * 1024];
  int index = 1;
  gnutls_datum_t g, n;
  gnutls_datum_t str_g, str_n;

  fd = fopen (tpasswd_conf, "w");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot open file '%s'\n", tpasswd_conf);
      return -1;
    }

  for (index = 1; index <= 5; index++)
    {

      if (index == 1)
        {
          n = gnutls_srp_1024_group_prime;
          g = gnutls_srp_1024_group_generator;
        }
      else if (index == 2)
        {
          n = gnutls_srp_1536_group_prime;
          g = gnutls_srp_1536_group_generator;
        }
      else if (index == 3)
        {
          n = gnutls_srp_2048_group_prime;
          g = gnutls_srp_2048_group_generator;
        }
      else if (index == 4)
        {
          n = gnutls_srp_3072_group_prime;
          g = gnutls_srp_3072_group_generator;
        }
      else if (index == 5)
        {
          n = gnutls_srp_4096_group_prime;
          g = gnutls_srp_4096_group_generator;
        }
      else
        {
          fprintf(stderr, "Unknown index: %d\n", index);
          return -1;
        }

      printf ("\nGroup %d, of %d bits:\n", index, n.size * 8);
      print_num ("Generator", &g);
      print_num ("Prime", &n);

      if (gnutls_srp_base64_encode_alloc (&n, &str_n) < 0)
        {
          fprintf (stderr, "Could not encode\n");
          return -1;
        }

      if (gnutls_srp_base64_encode_alloc (&g, &str_g) < 0)
        {
          fprintf (stderr, "Could not encode\n");
          return -1;
        }

      sprintf (line, "%d:%s:%s\n", index, str_n.data, str_g.data);

      gnutls_free (str_n.data);
      gnutls_free (str_g.data);

      fwrite (line, 1, strlen (line), fd);

    }

  fclose (fd);

  return 0;

}

/* The format of a tpasswd file is:
 * username:verifier:salt:index
 *
 * index is the index of the prime-generator pair in tpasswd.conf
 */
static int
_verify_passwd_int (const char *username, const char *passwd,
                    char *verifier, const char *salt,
                    const gnutls_datum_t * g, const gnutls_datum_t * n)
{
  char _salt[1024];
  gnutls_datum_t tmp, raw_salt, new_verifier;
  size_t salt_size;
  char *pos;

  if (salt == NULL || verifier == NULL)
    return -1;

  if (strlen(salt) >= sizeof(_salt))
    {
      fprintf (stderr, "Too long salt.\n");
      return -1;
    }

  /* copy salt, and null terminate after the ':' */
  strcpy (_salt, salt);
  pos = strchr (_salt, ':');
  if (pos != NULL)
    *pos = 0;

  /* convert salt to binary. */
  tmp.data = (void*)_salt;
  tmp.size = strlen (_salt);

  if (gnutls_srp_base64_decode_alloc (&tmp, &raw_salt) < 0)
    {
      fprintf (stderr, "Could not decode salt.\n");
      return -1;
    }

  if (gnutls_srp_verifier
      (username, passwd, &raw_salt, g, n, &new_verifier) < 0)
    {
      fprintf (stderr, "Could not make the verifier\n");
      return -1;
    }

  free (raw_salt.data);

  /* encode the verifier into _salt */
  salt_size = sizeof (_salt);
  memset (_salt, 0, salt_size);
  if (gnutls_srp_base64_encode (&new_verifier, _salt, &salt_size) < 0)
    {
      fprintf (stderr, "Encoding error\n");
      return -1;
    }

  free (new_verifier.data);

  if (strncmp (verifier, _salt, strlen (_salt)) == 0)
    {
      fprintf (stderr, "Password verified\n");
      return 0;
    }
  else
    {
      fprintf (stderr, "Password does NOT match\n");
    }
  return -1;
}

static int
filecopy (const char *src, const char *dst)
{
  FILE *fd, *fd2;
  char line[5 * 1024];
  char *p;

  fd = fopen (dst, "w");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot open '%s' for write\n", dst);
      return -1;
    }

  fd2 = fopen (src, "r");
  if (fd2 == NULL)
    {
      /* empty file */
      fclose (fd);
      return 0;
    }

  line[sizeof (line) - 1] = 0;
  do
    {
      p = fgets (line, sizeof (line) - 1, fd2);
      if (p == NULL)
        break;

      fputs (line, fd);
    }
  while (1);

  fclose (fd);
  fclose (fd2);

  return 0;
}

/* accepts password file */
static int
find_strchr (const char *username, const char *file)
{
  FILE *fd;
  char *pos;
  char line[5 * 1024];
  unsigned int i;

  fd = fopen (file, "r");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot open file '%s'\n", file);
      return -1;
    }

  while (fgets (line, sizeof (line), fd) != NULL)
    {
      /* move to first ':' */
      i = 0;
      while ((line[i] != ':') && (line[i] != '\0') && (i < sizeof (line)))
        {
          i++;
        }
      if (strncmp (username, line, MAX (i, strlen (username))) == 0)
        {
          /* find the index */
          pos = strrchr (line, ':');
          pos++;
          fclose (fd);
          return atoi (pos);
        }
    }

  fclose (fd);
  return -1;
}

/* Parses the tpasswd files, in order to verify the given
 * username/password pair.
 */
static int
verify_passwd (const char *conffile, const char *tpasswd, 
               const char *username, const char *passwd)
{
  FILE *fd;
  char line[5 * 1024];
  unsigned int i;
  gnutls_datum_t g, n;
  int iindex;
  char *p, *pos;

  iindex = find_strchr (username, tpasswd);
  if (iindex == -1)
    {
      fprintf (stderr, "Cannot find '%s' in %s\n", username, tpasswd);
      return -1;
    }

  fd = fopen (conffile, "r");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot find %s\n", conffile);
      return -1;
    }

  do
    {
      p = fgets (line, sizeof (line) - 1, fd);
    }
  while (p != NULL && atoi (p) != iindex);

  if (p == NULL)
    {
      fprintf (stderr, "Cannot find entry in %s\n", conffile);
      return -1;
    }
  line[sizeof (line) - 1] = 0;

  fclose (fd);

  if ((iindex = read_conf_values (&g, &n, line)) < 0)
    {
      fprintf (stderr, "Cannot parse conf file '%s'\n", conffile);
      return -1;
    }

  fd = fopen (tpasswd, "r");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot open file '%s'\n", tpasswd);
      return -1;
    }

  while (fgets (line, sizeof (line), fd) != NULL)
    {
      /* move to first ':' 
       * This is the actual verifier.
       */
      i = 0;
      while ((line[i] != ':') && (line[i] != '\0') && (i < sizeof (line)))
        {
          i++;
        }
      if (strncmp (username, line, MAX (i, strlen (username))) == 0)
        {
          char *verifier_pos, *salt_pos;

          pos = strchr (line, ':');
          fclose (fd);
          if (pos == NULL)
            {
              fprintf (stderr, "Cannot parse conf file '%s'\n", conffile);
              return -1;
            }
          pos++;
          verifier_pos = pos;

          /* Move to the salt */
          pos = strchr (pos, ':');
          if (pos == NULL)
            {
              fprintf (stderr, "Cannot parse conf file '%s'\n", conffile);
              return -1;
            }
          pos++;
          salt_pos = pos;

          return _verify_passwd_int (username, passwd,
                                     verifier_pos, salt_pos, &g, &n);
        }
    }

  fclose (fd);
  return -1;

}

#define KPASSWD "/etc/tpasswd"
#define KPASSWD_CONF "/etc/tpasswd.conf"

static void
tls_log_func (int level, const char *str)
{
  fprintf (stderr, "|<%d>| %s", level, str);
}

int main (int argc, char **argv)
{
  const char *passwd;
  int salt_size, ret;
  int optct;
  const char* fpasswd, *fpasswd_conf;
  const char* username;
#ifndef _WIN32
   struct passwd *pwd;
#endif

  set_program_name (argv[0]);

  if ((ret = gnutls_global_init ()) < 0)
    {
      fprintf (stderr, "global_init: %s\n", gnutls_strerror (ret));
      exit (1);
    }

  umask (066);

  optct = optionProcess( &srptoolOptions, argc, argv);
  argc -= optct;
  argv += optct;

  gnutls_global_set_log_function (tls_log_func);
  gnutls_global_set_log_level (OPT_VALUE_DEBUG);

  if (HAVE_OPT(CREATE_CONF))
    {
      return generate_create_conf (OPT_ARG(CREATE_CONF));
    }

  if (HAVE_OPT(PASSWD))
    fpasswd = OPT_ARG(PASSWD);
  else
    fpasswd = (char *) KPASSWD;

  if (HAVE_OPT(PASSWD_CONF))
    fpasswd_conf = OPT_ARG(PASSWD_CONF);
  else
    fpasswd_conf = (char *) KPASSWD_CONF;

  if (HAVE_OPT(USERNAME))
    username = OPT_ARG(USERNAME);
  else
    {
#ifndef _WIN32
      pwd = getpwuid (getuid ());

      if (pwd == NULL)
        {
          fprintf (stderr, "No such user\n");
          return -1;
        }

      username = pwd->pw_name;
#else
      fprintf (stderr, "Please specify a user\n");
      return -1;
#endif
    }

  salt_size = 16;

  passwd = getpass ("Enter password: ");
  if (passwd == NULL)
    {
      fprintf (stderr, "Please specify a password\n");
      return -1;
    }

/* not ready yet */
  if (HAVE_OPT(VERIFY))
    {
      return verify_passwd (fpasswd_conf, fpasswd,
                            username, passwd);
    }


  return crypt_int (username, passwd, salt_size,
                    fpasswd_conf, fpasswd, VALUE_OPT_INDEX);

}

static char *
_srp_crypt (const char *username, const char *passwd, int salt_size,
            const gnutls_datum_t * g, const gnutls_datum_t * n)
{
  unsigned char salt[128];
  static char result[1024];
  gnutls_datum_t dat_salt, txt_salt;
  gnutls_datum_t verifier, txt_verifier;

  if ((unsigned) salt_size > sizeof (salt))
    return NULL;

  /* generate the salt
   */
  if (gnutls_rnd (GNUTLS_RND_NONCE, salt, salt_size) < 0)
    {
      fprintf (stderr, "Could not create nonce\n");
      return NULL;
    }

  dat_salt.data = salt;
  dat_salt.size = salt_size;

  if (gnutls_srp_verifier (username, passwd, &dat_salt, g, n, &verifier) < 0)
    {
      fprintf (stderr, "Error getting verifier\n");
      return NULL;
    }

  /* base64 encode the verifier */
  if (gnutls_srp_base64_encode_alloc (&verifier, &txt_verifier) < 0)
    {
      fprintf (stderr, "Error encoding\n");
      free (verifier.data);
      return NULL;
    }

  free (verifier.data);

  if (gnutls_srp_base64_encode_alloc (&dat_salt, &txt_salt) < 0)
    {
      fprintf (stderr, "Error encoding\n");
      return NULL;
    }

  sprintf (result, "%s:%s", txt_verifier.data, txt_salt.data);
  free (txt_salt.data);
  free (txt_verifier.data);

  return result;

}


int
crypt_int (const char *username, const char *passwd, int salt_size,
           const char *tpasswd_conf, const char *tpasswd, int uindex)
{
  FILE *fd;
  char *cr;
  gnutls_datum_t g, n;
  char line[5 * 1024];
  char *p, *pp;
  int iindex;
  char tmpname[1024];

  fd = fopen (tpasswd_conf, "r");
  if (fd == NULL)
    {
      fprintf (stderr, "Cannot find %s\n", tpasswd_conf);
      return -1;
    }

  do
    {                           /* find the specified uindex in file */
      p = fgets (line, sizeof (line) - 1, fd);
      iindex = atoi (p);
    }
  while (p != NULL && iindex != uindex);

  if (p == NULL)
    {
      fprintf (stderr, "Cannot find entry in %s\n", tpasswd_conf);
      return -1;
    }
  line[sizeof (line) - 1] = 0;

  fclose (fd);
  if ((iindex = read_conf_values (&g, &n, line)) < 0)
    {
      fprintf (stderr, "Cannot parse conf file '%s'\n", tpasswd_conf);
      return -1;
    }

  cr = _srp_crypt (username, passwd, salt_size, &g, &n);
  if (cr == NULL)
    {
      fprintf (stderr, "Cannot _srp_crypt()...\n");
      return -1;
    }
  else
    {
      /* delete previous entry */
      struct stat st;
      FILE *fd2;
      int put;

      if (strlen (tpasswd) > sizeof (tmpname) + 5)
        {
          fprintf (stderr, "file '%s' is tooooo long\n", tpasswd);
          return -1;
        }
      strcpy (tmpname, tpasswd);
      strcat (tmpname, ".tmp");

      if (stat (tmpname, &st) != -1)
        {
          fprintf (stderr, "file '%s' is locked\n", tpasswd);
          return -1;
        }

      if (filecopy (tpasswd, tmpname) != 0)
        {
          fprintf (stderr, "Cannot copy '%s' to '%s'\n", tpasswd, tmpname);
          return -1;
        }

      fd = fopen (tpasswd, "w");
      if (fd == NULL)
        {
          fprintf (stderr, "Cannot open '%s' for write\n", tpasswd);
          remove (tmpname);
          return -1;
        }

      fd2 = fopen (tmpname, "r");
      if (fd2 == NULL)
        {
          fprintf (stderr, "Cannot open '%s' for read\n", tmpname);
          remove (tmpname);
          return -1;
        }

      put = 0;
      do
        {
          p = fgets (line, sizeof (line) - 1, fd2);
          if (p == NULL)
            break;

          pp = strchr (line, ':');
          if (pp == NULL)
            continue;

          if (strncmp (p, username,
                       MAX (strlen (username), (unsigned int) (pp - p))) == 0)
            {
              put = 1;
              fprintf (fd, "%s:%s:%u\n", username, cr, iindex);
            }
          else
            {
              fputs (line, fd);
            }
        }
      while (1);

      if (put == 0)
        {
          fprintf (fd, "%s:%s:%u\n", username, cr, iindex);
        }

      fclose (fd);
      fclose (fd2);

      remove (tmpname);

    }


  return 0;
}



/* this function parses tpasswd.conf file. Format is:
 * int(index):base64(n):base64(g)
 */
static int
read_conf_values (gnutls_datum_t * g, gnutls_datum_t * n, char *str)
{
  char *p;
  int len;
  int index, ret;
  gnutls_datum_t dat;

  index = atoi (str);

  p = strrchr (str, ':');       /* we have g */
  if (p == NULL)
    {
      return -1;
    }

  *p = '\0';
  p++;

  /* read the generator */
  len = strlen (p);
  if (p[len - 1] == '\n')
    len--;

  dat.data = (void*)p;
  dat.size = len;
  ret = gnutls_srp_base64_decode_alloc (&dat, g);

  if (ret < 0)
    {
      fprintf (stderr, "Decoding error\n");
      return -1;
    }

  /* now go for n - modulo */
  p = strrchr (str, ':');       /* we have n */
  if (p == NULL)
    {
      return -1;
    }

  *p = '\0';
  p++;

  dat.data = (void*)p;
  dat.size = strlen (p);

  ret = gnutls_srp_base64_decode_alloc (&dat, n);

  if (ret < 0)
    {
      fprintf (stderr, "Decoding error\n");
      free (g->data);
      return -1;
    }

  return index;
}

extern void srptool_version (void);

void
srptool_version (void)
{
  const char *p = PACKAGE_NAME;
  if (strcmp (gnutls_check_version (NULL), PACKAGE_VERSION) != 0)
    p = PACKAGE_STRING;
  version_etc (stdout, "srptool", p, gnutls_check_version (NULL),
               "Nikos Mavrogiannopoulos", (char *) NULL);
}
